bip-raw-as-subscript.md

```diff
@@ -0,0 +1,101 @@
+<pre>
+  BIP: ???
+  Layer: Applications
+  Title: raw() as subscript in Output Script Descriptors
+  Author: Matias Furszyfer <mfurszy@protonmail.com>
+  Comments-Summary: No comments yet.
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-???
+  Status: Draft
+  Type: Standards Track
+  Created: ?
+  License: BSD-2-Clause
+</pre>
+
+==Abstract==
+
+This document specifies `raw()` as subscript for output script descriptors.
+`raw()` encapsulates a raw hex script. This BIP allows `raw()` to be used in
+the context of other descriptors.
+
+==Copyright==
+
+This BIP is licensed under the BSD 2-clause license.
+
+==Motivation==
+
+Allowing arbitrary hex data to be wrapped in `sh()`, `wsh()`, or even within the `TREE`
+argument of a `tr(KEY, TREE)` descriptor enables the representation of currently
+inexpressible information in the descriptors' language.
+
+Specifically, the absence of this feature limits the representation of non-standard redeem
+and witness scripts. This occurs because they can currently only be represented as top-level
+`raw(HEX)` descriptors, which retain only the output script information and lack the ability
+to preserve the actual script.
+
+Additionally, as noted [here](https://github.com/bitcoin/bitcoin/issues/24114#issuecomment-1127978154),
+there are other useful scenarios for this feature. For example, it allows representing
+in a descriptor that we lack complete knowledge of all solvability paths but can still
+solve the output. This includes cases like a taproot tree where we know only one of its
+paths. Or, participating in signing a script without knowing all participants' keys,
+relying solely on the script structure.
+
+==Specification==
+
+### For `sh()` and `wsh()` descriptors:
+`raw(HEX)` must represent the arbitrary script data within the provided context.
+This for example means that a P2SH output script (OP_HASH160 <hash160(HEX_script)> OP_EQUAL)
+must be created from the provided hex data when a `sh(raw(HEX))` descriptor is provided.
+
+Parallelly, a P2WSH output script (OP_0 <hash160(HEX_script)>) must be created from the provided
+hex data when a `wsh(raw(HEX))` descriptor is provided.
+
+### For `tr(KEY, TREE)` descriptors:
+Two new fragments are allowed within the taproot `TREE` context: `rawnode(HEX)` and `raw(HEX)`.
+
+#### `rawnode(HEX)`:
+Indicating a tree node with specified 32-bit hash, but no specified subtree.
+This can serve as either a tree branch or the root of the Merkle tree.
+
+#### `raw(HEX)`:
+Defines a tree leaf containing the specified script in hex.
+Note: The leaf version must be internally fixed at the existing `0xC0` to prevent introducing
+unsupported or undefined functionality.
+If a different version is required for any use case, a new BIP could introduce `raw(HEX, VERSION)`
+in the future.
```

Comment:
I think we should just support `raw(HEX, VERSION)` and be forward compatible which goes in line with what was suggested in the https://github.com/bitcoin/bitcoin/issues/24114